### PR TITLE
Make remote config accessible to background agents

### PR DESCRIPTION
**Pull request by SlayterDev**
Task/Issue URL: https://app.asana.com/0/1203581873609357/1207165680693234/f
Tech Design URL:
CC:

**Description**:

**Steps to test this PR**:
1. See https://github.com/duckduckgo/BrowserServicesKit/pull/947
2.

**Definition of Done (Internal Only)**:

* [ ] Does this PR satisfy our [Definition of Done](https://app.asana.com/0/1202500774821704/1207634633537039/f)?

**Copy Testing**:

* [ ] Use of correct apostrophes in new copy, ie `’` rather than `'`

**Orientation Testing**:

* [ ] Portrait
* [ ] Landscape

**Device Testing**:

* [ ] iPhone SE (1st Gen)
* [ ] iPhone 8
* [ ] iPhone X
* [ ] iPhone 14 Pro
* [ ] iPad

**OS Testing**:

* [ ] iOS 15
* [ ] iOS 16
* [ ] iOS 17

**Theme Testing**:

* [ ] Light theme
* [ ] Dark theme

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

### CODE REVIEW

File: Core/FileStore.swift
Comment (samsymons) [inline Core/FileStore.swift]: As we discussed, this can be updated to use a coordinator-specific pixel

File: Core/FileStore.swift
Comment (samsymons) [inline Core/FileStore.swift:34]: Let's check that this `persist` function overall is never called on the main thread, since the `coordinate` function will block that thread until it can complete what it's trying to do.

File: Core/PixelEvent.swift
Comment (samsymons) [inline Core/PixelEvent.swift]: Let's update this to add the `Configuration` name as a suffix.

File: DuckDuckGo/AppDelegate.swift
Comment (samsymons) [inline DuckDuckGo/AppDelegate.swift]: Since this used to be a singleton (and I agree it shouldn't be any more), we can probably play it safe and put an instance of it on the `AppDependencyProvider`, just to make sure there are no assumptions that other parts of the codebase are making.

File: DuckDuckGo/Configuration/ConfigurationManager.swift
Comment (samsymons) [inline DuckDuckGo/Configuration/ConfigurationManager.swift]: Per [file presenter documentation](https://developer.apple.com/documentation/foundation/nsfilecoordinator), we can use `willEnterForegroundNotification` here instead.

File: PacketTunnelProvider/NetworkProtection/ConfigurationStore.swift
Comment (samsymons) [inline PacketTunnelProvider/NetworkProtection/ConfigurationStore.swift:35]: This key can likely be updated to mention that it's an ETag:

```suggestion
        return "privacyConfigurationEtag"
```

File: PacketTunnelProvider/NetworkProtection/NetworkProtectionPacketTunnelProvider.swift
Comment (samsymons) [inline PacketTunnelProvider/NetworkProtection/NetworkProtectionPacketTunnelProvider.swift]:
```suggestion
        configurationManager = ConfigurationManager()
```

File: PacketTunnelProvider/NetworkProtection/NetworkProtectionPacketTunnelProvider.swift
Comment (samsymons) [inline PacketTunnelProvider/NetworkProtection/NetworkProtectionPacketTunnelProvider.swift]: Since this is in the init function of the VPN, it won't necessarily react to Privacy Config changes quickly - the VPN is intended to run indefinitely, so we'll want to instead observe config refreshes and fire it there.

We may be able to use the `UniquePixel` API as well, to save some UserDefaults boilerplate.

Comment (samsymons) [reply PacketTunnelProvider/NetworkProtection/NetworkProtectionPacketTunnelProvider.swift]: I'd actually consider using the `DailyPixel` API here instead since it'll give us an accurate look each day at the exact rollout of the flag.

File: PacketTunnelProvider/NetworkProtection/ConfigurationStore.swift
Comment (SlayterDev) [reply PacketTunnelProvider/NetworkProtection/ConfigurationStore.swift:35]: This was set to match the main app which has more etags then just remote config. None of those keys mention the etag so if we update this one we should update them all in the main app.

File: Core/FileStore.swift
Comment (SlayterDev) [reply Core/FileStore.swift:34]: I tracked the persist calls all the way up to [here](https://github.com/duckduckgo/iOS/blob/1c736e9c4d9befca0a42ed280b4d47c2a5f38800/DuckDuckGo/AppConfigurationFetch.swift#L172). Basically the full fetch then store process is initiated by the `AppConfigurationFetch` within a `Task`. AFAIK that should put it off the main thread.

File: PacketTunnelProvider/NetworkProtection/NetworkProtectionPacketTunnelProvider.swift
Comment (samsymons) [inline PacketTunnelProvider/NetworkProtection/NetworkProtectionPacketTunnelProvider.swift:342]: I think it's possible we'll still miss calls to this pixel - this publisher is only called when we receive updates, so if no Privacy Config update is published on a given day then we'll get no events coming through.

We could probably add the `BackgroundAgentPixelTestSubfeature.pixelTest` check and pixel call somewhere that we _know_ will get called regularly, like `handleConnectionStatusChange`. That said, we only really care about getting these calls on days where we're changing the rollout param and at that point we _will_ get this publisher called, so I'd be fine with it staying as-is.

File: PacketTunnelProvider/NetworkProtection/VPNPrivacyConfigurationManager.swift
Comment (samsymons) [inline PacketTunnelProvider/NetworkProtection/VPNPrivacyConfigurationManager.swift:49]: I think we should close this gap as quickly as possible; from my perspective we can live with this in the short term, at least as part of running the pixel rollout test, but I'd advocate for us getting real embedded data working ASAP. Would you be able to scope the changes required for this for both iOS and macOS?

File: PacketTunnelProvider/NetworkProtection/ConfigurationStore.swift
Comment (samsymons) [reply PacketTunnelProvider/NetworkProtection/ConfigurationStore.swift:35]: No worries, I think we can leave this one in that case - thanks!

File: PacketTunnelProvider/NetworkProtection/VPNPrivacyConfigurationManager.swift
Comment (samsymons) [inline PacketTunnelProvider/NetworkProtection/VPNPrivacyConfigurationManager.swift:1]: This class could use unit tests, though I acknowledge there's no good test suite set up for it right now. I'll defer to @diegoreymendez on whether he'd like to block on this, as primary of the VPN AOR.

**Review by samsymons — APPROVED** [review]
Approving based on the behavior working well for me across a bunch of different test cases.

File: PacketTunnelProvider/NetworkProtection/VPNPrivacyConfigurationManager.swift
Comment (SlayterDev) [reply PacketTunnelProvider/NetworkProtection/VPNPrivacyConfigurationManager.swift:49]: Follow up opened here: https://app.asana.com/0/1200890834746050/1208305934034181/f

File: Core/EtagStorage.swift
Comment (diegoreymendez) [inline Core/EtagStorage.swift:34]: Is this a risky transition / migration?  We're moving from using an existing user defaults suite, to using a new one.

Is there anything stored here that will get reset that might cause trouble for users?

File: Core/FileStore.swift
Comment (diegoreymendez) [inline Core/FileStore.swift:25]: Since this is used for the file paths: is there anything that will get reset here that might cause trouble for users?

File: Core/UserDefaults+NetworkProtection.swift
Comment (diegoreymendez) [inline Core/UserDefaults+NetworkProtection.swift]: Can this be replaced by `ContentBlockerStoreConstants. configurationGroupName`?

File: DuckDuckGo/AppDelegate.swift
Comment (diegoreymendez) [inline DuckDuckGo/AppDelegate.swift:522]: Should this happen before `AppDependencyProvider.shared.configurationManager` is used in line 295?

File: PacketTunnelProvider/NetworkProtection/VPNPrivacyConfigurationManager.swift
Comment (diegoreymendez) [reply PacketTunnelProvider/NetworkProtection/VPNPrivacyConfigurationManager.swift:1]: Agree that it'd be ideal to have tests here, but I'm not confident the effort to implement those tests is worth additional delay here.  Especially because the effort involved isn't completely clear to me and I don't want to be blocking things here.

File: PacketTunnelProvider/NetworkProtection/NetworkProtectionPacketTunnelProvider.swift
Comment (diegoreymendez) [reply PacketTunnelProvider/NetworkProtection/NetworkProtectionPacketTunnelProvider.swift:342]: Struggling a bit to follow what `pixelTest` is.

**Review by diegoreymendez — COMMENTED** [review]
Added some comments.  Mostly questions I have.

File: Core/EtagStorage.swift
Comment (SlayterDev) [reply Core/EtagStorage.swift:34]: I don't think there's a risk here. Configuration starts up and fetches all configs at app launch. For an app update this will look like a fresh install for the configs. It will see no etags and just fetch the current versions and store them. I haven't run in to any issues during testing which involved a "migration".

File: Core/FileStore.swift
Comment (SlayterDev) [reply Core/FileStore.swift:25]: Same as the etags. The module will see no files and download fresh ones. In the case of content blocking/privacy config we have embedded fallbacks while we wait for downloads.

File: DuckDuckGo/AppDelegate.swift
Comment (SlayterDev) [reply DuckDuckGo/AppDelegate.swift:522]: It doesn't matter. We don't use the built in timer of the configuration manager because of background app refresh. And even if we did the only purpose of this method is to diff the stored etag vs the etag in the privacy config manager and load fresh from disk if needed.

**Review by diegoreymendez — APPROVED** [review]
Nothing more to add, my main doubts have been covered here.

**Pull request merged**
